### ISSUE TITLE
problem running MPI under Flux

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#220]
Following up on this report from Luke Johntson on flux-discuss:

> An odd thing I’ve noticed is that if I don’t do any of the above before trying to run with MPI, then wreckrun works with the first MPI application I run for each new flux instance, even though it appears that PMI is not used, as `flux kvs dir pmi` doesn’t find anything. After that first success, all subsequent MPI applications fail with the error:
```
srun: mvapich: 2015-06-17T11:03:28: ABORT from MPI rank 0 [on rzmerl3] dest rank 9 [on (null)]
PMGR_COLLECTIVE ERROR: rank 0 on rzmerl3: Reading from mpirun at
192.168.64.156:39298 (read(buf=60a300,size=12) Success errno=0) @ file
pmgr_collective_client_mpirun.c:60
PMGR_COLLECTIVE ERROR: rank 0 on rzmerl3: Nonblocking connect failed immediately
connecting to 0.0.232.142:43828 (connect() Invalid argument errno=22) @ file
pmgr_collective_client_common.c:86
srun: mvapich: 2015-06-17T11:03:28: ABORT from MPI rank 1 [on rzmerl4] dest rank
9 [on (null)]
Hangup
rzmerl156{23}:/g/g19/luketj> srun: error: rzmerl3: task 0: Killed
```

Building against current master (ad54d7d), having disabled the lc dotkit environment and run `module load mvapich-gnu-gen2` I find I can reproduce this with src/test/mpi_hello.  The first run works, the second one fails spectacularly, killing flux:
```
bash-4.1$ ./flux wreckrun -N4 ../test/mpi_hello
wreckrun: 0.002s: Sending LWJ request for 1 tasks (cmdline "../test/mpi_hello")
wreckrun: 0.008s: Registered jobid 7
wreckrun: Allocating 4 tasks across 4 available nodes..
wreckrun: tasks per node: node[0-3]: 1
wreckrun: 0.010s: Sending run event
wreckrun: 0.013s: State = reserved
wreckrun: 0.035s: State = starting
wreckrun: 0.047s: State = running
0: completed first barrier
0: completed MPI_Init in 1.759s.  There are 4 tasks
0: completed MPI_Finalize
wreckrun: 2.324s: State = complete
wreckrun: All tasks completed successfully.
bash-4.1$ ./flux wreckrun -N4 ../test/mpi_hello
wreckrun: 0.002s: Sending LWJ request for 1 tasks (cmdline "../test/mpi_hello")
wreckrun: 0.008s: Registered jobid 8
wreckrun: Allocating 4 tasks across 4 available nodes..
wreckrun: tasks per node: node[0-3]: 1
wreckrun: 0.010s: Sending run event
wreckrun: 0.013s: State = reserved
wreckrun: 0.035s: State = starting
wreckrun: 0.049s: State = running
srun: mvapich: 2015-06-17T12:05:56: ABORT from MPI rank 0 [on hype203] dest rank 9 [on (null)]
              PMGR_COLLECTIVE ERROR: rank 0 on hype203: Reading from mpirun at 192.168.81.157:33084 (read(buf=612280,size=24) Success errno=0) @ file pmgr_collective_client_mpirun.c:60
srun: mvapich: 2015-06-17T12:05:56: ABORT from MPI rank 2 [on hype235] dest rank 9 [on (null)]
              Hangup
bash-4.1$ srun: mvapich: 2015-06-17T12:05:56: ABORT from MPI rank 1 [on hype232] dest rank 9 [on (null)]
                        srun: mvapich: 2015-06-17T12:05:56: ABORT from MPI rank 3 [on hype236] dest rank 9 [on (null)]
                                      srun: error: hype203: task 0: Killed
                                                                          srun: First task exited 30s ago
                         srun: tasks 1-3: running
                                                 srun: task 0: exited abnormally
                                                                               srun: Terminating job step 1112374.0
                                   srun: Job step aborted: Waiting up to 2 seconds for job step to finish.
                          srun: error: Timed out waiting for job step to complete
```
On the other hand, running `./flux wreckrun -N4 hostname` multiple times works as expected.

**Comment by garlick** [comment]
Oops my attempt to confirm was neither using mvapich2 nor the correct interconnect version for hype, where I was testing.  After getting that correct, reconfiguring flux, and rebuilding mpi_hello, things are working properly with the mpi_hello program
```
hype356@garlick:./flux start -N4
srun: Job is in held state, pending scheduler release
srun: job 1112379 queued and waiting for resources
srun: job 1112379 has been allocated resources
flux: [1434569811.849151] broker.info[0] nodeset: [0-3] (complete)
flux: [1434569811.849215] broker.info[0] starting shell
bash-4.1$ ./flux wreckrun -N 4 hostname
wreckrun: 0.002s: Sending LWJ request for 1 tasks (cmdline "hostname")
wreckrun: 0.009s: Registered jobid 1
wreckrun: Allocating 4 tasks across 4 available nodes..
wreckrun: tasks per node: node[0-3]: 1
wreckrun: 0.015s: Sending run event
wreckrun: 0.018s: State = reserved
wreckrun: 0.049s: State = starting
wreckrun: 0.095s: State = running
hype232
hype235
hype203
hype236
wreckrun: 0.113s: State = complete
wreckrun: All tasks completed successfully.
bash-4.1$ ^C
bash-4.1$ aset -o vi
bash: aset: command not found
bash-4.1$ set -o vi
bash-4.1$ ./flux wreckrun -N 4 ../test/mpi_hello
wreckrun: 0.002s: Sending LWJ request for 1 tasks (cmdline "../test/mpi_hello")
wreckrun: 0.007s: Registered jobid 2
wreckrun: Allocating 4 tasks across 4 available nodes..
wreckrun: tasks per node: node[0-3]: 1
wreckrun: 0.008s: Sending run event
wreckrun: 0.011s: State = reserved
wreckrun: 0.034s: State = starting
wreckrun: 0.050s: State = running
0: completed MPI_Init in 0.044s.  There are 4 tasks
0: completed first barrier
0: completed MPI_Finalize
wreckrun: 2.365s: State = complete
wreckrun: All tasks completed successfully.
bash-4.1$ ./flux wreckrun -N 4 ../test/mpi_hello
wreckrun: 0.002s: Sending LWJ request for 1 tasks (cmdline "../test/mpi_hello")
wreckrun: 0.008s: Registered jobid 3
wreckrun: Allocating 4 tasks across 4 available nodes..
wreckrun: tasks per node: node[0-3]: 1
wreckrun: 0.010s: Sending run event
wreckrun: 0.013s: State = reserved
wreckrun: 0.036s: State = starting
wreckrun: 0.050s: State = running
0: completed MPI_Init in 0.051s.  There are 4 tasks
0: completed first barrier
0: completed MPI_Finalize
wreckrun: 2.231s: State = complete
wreckrun: All tasks completed successfully.
bash-4.1$ ./flux wreckrun -N 4 ../test/mpi_hello
wreckrun: 0.002s: Sending LWJ request for 1 tasks (cmdline "../test/mpi_hello")
wreckrun: 0.008s: Registered jobid 4
wreckrun: Allocating 4 tasks across 4 available nodes..
wreckrun: tasks per node: node[0-3]: 1
wreckrun: 0.011s: Sending run event
wreckrun: 0.014s: State = reserved
wreckrun: 0.034s: State = starting
wreckrun: 0.048s: State = running
0: completed first barrier
0: completed MPI_Init in 0.049s.  There are 4 tasks
0: completed MPI_Finalize
wreckrun: 2.227s: State = complete
wreckrun: All tasks completed successfully.
bash-4.1$ module list
Currently Loaded Modulefiles:
  1) mvapich2-gnu-psm/1.7
```

Luke, could you verify that you've got the right mvapich2 interconnect type loaded for the machine you're testing on?

**Comment by grondo** [comment]
In the first example it appears the mpi being used is plain mvapich (no PMI) which expects to connect back to srun. I'd guess the first launch works because it happens to connect back to the srun which launched flux (maybe more environment scrubbing needed in `flux-start`?)

**Comment by garlick** [comment]
Ah, that makes sense, mvapich convinces srun that job step 0 aborted, so flux is killed.  This apparently helps:
```
bash-4.1$ unset MPIRUN_HOST
bash-4.1$ ./flux wreckrun -N4 ../test/mpi_hello
wreckrun: 0.002s: Sending LWJ request for 1 tasks (cmdline "../test/mpi_hello")
wreckrun: 0.008s: Registered jobid 6
wreckrun: Allocating 4 tasks across 4 available nodes..
wreckrun: tasks per node: node[0-3]: 1
wreckrun: 0.012s: Sending run event
wreckrun: 0.015s: State = reserved
wreckrun: 0.040s: State = starting
wreckrun: 0.153s: State = running
PMGR_COLLECTIVE ERROR: rank 0 on hype228: Missing required environment variable: MPIRUN_HOST @ pmgr_collective_common.c:115
PMGR_COLLECTIVE ERROR: rank 1 on hype229: Missing required environment variable: MPIRUN_HOST @ pmgr_collective_common.c:115
PMGR_COLLECTIVE ERROR: rank 3 on hype231: Missing required environment variable: MPIRUN_HOST @ pmgr_collective_common.c:115
PMGR_COLLECTIVE ERROR: rank 2 on hype230: Missing required environment variable: MPIRUN_HOST @ pmgr_collective_common.c:115
wreckrun: 0.177s: State = complete
wreckrun: tasks [0-3]: exited with exit code 1
```

**Comment by grondo** [comment]
We should probably sanitize all MPIRUN* environment variables (and maybe others) in `flux-start`? (At least I can't think of a case where we'd want to propagate those down to initial program of a flux instance, even with flux-within-flux?)

**Comment by garlick** [comment]
`flux-start` calls srun, and then slurm starts flux-broker, so I guess the environment would need to be cleaned at flux-broker or somewhere before anything else gets executed by it.

**Comment by grondo** [comment]
@garlick, Duh, sorry. Yeah that is what I meant.

**Comment by grondo** [comment]
Oh, one other thing we could try is adding `--mpi=none` to `srun` flags. I don't think this disables PMI, but it will disable mvapich/pmgr support.

**Comment by garlick** [comment]
Yes that does the trick (fast failure for wrong MPI, works for right MPI).

**Comment by luketjohnston** [comment]
Excellent, I've got it working now. I don't know how to verify that I have the right mvapich2 interconnect type for my machine (rzmerl), but I just did "module load mvapich2-gnu-psm/1.7" based on your output above, and reconfigured flux, and mpi_hello works now. Thanks! I haven't used the module command before, so I should probably look into what exactly that does.

**Comment by grondo** [comment]
Not to divert us too much -- and I don't know what the lifetime of PMGR-based MPI bootstrap is -- but it would be interesting to add support for another MPI initialization protocol besides PMI as an experiment -- at some point. I think Adam repackaged PMGR into a standalone library at some point in the past?

**Comment by dongahn** [comment]
As part of my early work, I actually have a PMGR port on top of our PMI. This can be support probably with minor modification... (In an all-day meeting today, I can take a look at this when I come back)

**Comment by garlick** [comment]
@luketjohnston sorry for the fragile stuff you tripped over in the process.  I think if mpi_hello works you have the correct runes to move forward.

**Comment by grondo** [comment]
@dongahn: Cool, that would be neat.

**Comment by luketjohnston** [comment]
@garlick No problem! I also just noticed that this doesn't work for the rzzeus machine, so I may have to figure that out down the road, but it's working for rzmerl and yeah that's good enough to move forward. Thanks for your help!

**Comment by grondo** [comment]
@luketjohnston rzzeus machine uses mlx adapters so you'd need use the `ofa` version of mvapich2:
```
module load mvapich2-gnu-ofa/1.7 
```

**Comment by luketjohnston** [comment]
@grondo Ok, thanks. Is there a reference of some sort I can use to determine what version of mvapich2 to use if I want to use it on other machines? Although rzmerl and rzzeus will probably be plenty for my purposes for now.

**Comment by garlick** [comment]
[this table](https://computing.llnl.gov/?set=resources&page=OCF_resources) might be useful?

**Comment by garlick** [comment]
Closing this issue as we merged #222, and captured other issues to solve in #221 and #223.
